### PR TITLE
Automatically account lockout after some attempts to login with wrong password

### LINKED ISSUE #11979: [FSTEC] Blocking users when the number of authentication attempts is exceeded
KIKIMR-21863

### PR DESCRIPTION
**Pull request by molotkov-and**
### Changelog entry 

Automatically account lockout after some attempts to login with wrong password.
User is unlocked after `AttemptResetDuration` time in config. If  `AttemptResetDuration` is 0 then user is locked out permanently

### Changelog category 

* New feature

### Additional information

...

### CODE REVIEW

File: ydb/core/tx/schemeshard/ut_login/ut_login.cpp [hunk 1]
```diff
@@ -177,6 +192,171 @@ Y_UNIT_TEST_SUITE(TSchemeShardLoginTest) {
...
+        resultLogin = Login(runtime, "user1", "wrongpassword5");
+        UNIT_ASSERT_VALUES_EQUAL(resultLogin.error(), TStringBuilder() << "User user1 is locked out");
+
+        runtime.SimulateSleep(TDuration::Seconds(7));
+
+        resultLogin = Login(runtime, "user1", "wrongpassword6");
+        UNIT_ASSERT_VALUES_EQUAL(resultLogin.error(), "Invalid password");
+        resultLogin = Login(runtime, "user1", "password1");
+        UNIT_ASSERT_VALUES_EQUAL(resultLogin.error(), "");
+        auto describe = DescribePath(runtime, TTestTxConfig::SchemeShard, "/MyRoot");
```
Comment (azevaykin) [inline ydb/core/tx/schemeshard/ut_login/ut_login.cpp:224]: This code is repeated in other tests. Let's move it to the function.

File: ydb/core/tx/schemeshard/ut_login/ut_login.cpp [hunk 2]
```diff
@@ -177,6 +192,171 @@ Y_UNIT_TEST_SUITE(TSchemeShardLoginTest) {
...
+        runtime.AddAppDataInit([](ui32 nodeIdx, NKikimr::TAppData& appData) {
+            Y_UNUSED(nodeIdx);
+            auto accountLockout = appData.AuthConfig.MutableAccountLockout();
+            // AttemptThreshold = 4
+            accountLockout->SetAttemptResetDuration("5s");
+        });
+        TTestEnv env(runtime);
+        ui64 txId = 100;
+        CreateAlterLoginCreateUser(runtime, ++txId, "/MyRoot", "user1", "password1");
+        auto resultLogin = Login(runtime, "user1", "wrongpassword1");
```
Comment (azevaykin) [inline ydb/core/tx/schemeshard/ut_login/ut_login.cpp:207]: May be `for(size_t attempt = 0 ; attempt++ ; attempt <=5 )` ?

File: ydb/library/login/login.h [hunk 3]
```diff
@@ -48,8 +48,16 @@ class TLoginProvider {
...
+        enum class EStatus {
+            SUCCESS,
+            INVALID_USER,
+            INVALID_PASSWORD,
+            UNAVAILABLE_KEY
+        };
+
         TString Token;
         TString SanitizedToken; // Token for audit logs
+        EStatus Status = EStatus::SUCCESS;
```
Comment (azevaykin) [inline ydb/library/login/login.h:60]: May be `EStatus::SUCCESS` is not a default value?

File: ydb/core/tx/schemeshard/schemeshard__login.cpp [hunk 4]
```diff
@@ -67,37 +69,131 @@ struct TSchemeShard::TTxLogin : TSchemeShard::TRwTxBase {
...
+
+    bool HandleExternalAuth(const NLogin::TLoginProvider::TLoginUserRequest& loginRequest) {
+        const NLogin::TLoginProvider::TLoginUserResponse loginResponse = Self->LoginProvider.LoginUser(loginRequest);
+        switch (loginResponse.Status) {
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::SUCCESS: {
+            SetSuccessResult(loginResponse);
+            break;
+        }
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::INVALID_PASSWORD:
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::INVALID_USER:
```
Comment (azevaykin) [inline ydb/core/tx/schemeshard/schemeshard__login.cpp:110]: Is it covered with unit tests?

File: ydb/core/tx/schemeshard/schemeshard__login.cpp [hunk 5]
```diff
@@ -67,37 +69,131 @@ struct TSchemeShard::TTxLogin : TSchemeShard::TRwTxBase {
...
+        Result->Record.SetSanitizedToken(loginResponse.SanitizedToken);
+    }
+
+    bool HandleLoginAuth(const NLogin::TLoginProvider::TLoginUserRequest& loginRequest, NIceDb::TNiceDb& db, const TActorContext& ctx) {
+        auto row = db.Table<Schema::LoginSids>().Key(loginRequest.User).Select();
+        if (!row.IsReady()) {
+            return false;
+        }
+        if (!row.IsValid()) {
+            Result->Record.SetError(TStringBuilder() << "Cannot find user: " << loginRequest.User);
```
Comment (azevaykin) [inline ydb/core/tx/schemeshard/schemeshard__login.cpp:130]: Is it covered with unit tests?

File: ydb/core/tx/schemeshard/schemeshard__login.cpp [hunk 6]
```diff
@@ -67,37 +69,131 @@ struct TSchemeShard::TTxLogin : TSchemeShard::TRwTxBase {
...
+    bool HandleExternalAuth(const NLogin::TLoginProvider::TLoginUserRequest& loginRequest) {
+        const NLogin::TLoginProvider::TLoginUserResponse loginResponse = Self->LoginProvider.LoginUser(loginRequest);
+        switch (loginResponse.Status) {
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::SUCCESS: {
+            SetSuccessResult(loginResponse);
+            break;
+        }
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::INVALID_PASSWORD:
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::INVALID_USER:
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::UNAVAILABLE_KEY: {
```
Comment (azevaykin) [inline ydb/core/tx/schemeshard/schemeshard__login.cpp:111]: Is it covered with unit tests?

File: ydb/core/tx/schemeshard/schemeshard_impl.cpp [hunk 7]
```diff
@@ -4413,6 +4413,20 @@ TActorId TSchemeShard::TPipeClientFactory::CreateClient(const TActorContext& ctx
...
+TSchemeShard::TAccountLockout::TAccountLockout(const ::NKikimrProto::TAccountLockout& accountLockout)
+    : AttemptThreshold(accountLockout.GetAttemptThreshold())
+{
+    AttemptResetDuration = TDuration::Zero();
+    if (accountLockout.GetAttemptResetDuration().empty()) {
+        return;
+    }
+    if (TDuration::TryParse(accountLockout.GetAttemptResetDuration(), AttemptResetDuration)) {
+        if (AttemptResetDuration.Seconds() == 0) {
+            AttemptResetDuration = TDuration::Zero();
```
Comment (azevaykin) [inline ydb/core/tx/schemeshard/schemeshard_impl.cpp:4425]: Is it covered with unit tests?

File: ydb/core/tx/schemeshard/ut_login/ut_login.cpp (hunk 1 above)
Comment (molotkov-and) [reply ydb/core/tx/schemeshard/ut_login/ut_login.cpp:224]: done

File: ydb/core/tx/schemeshard/ut_login/ut_login.cpp (hunk 2 above)
Comment (molotkov-and) [reply ydb/core/tx/schemeshard/ut_login/ut_login.cpp:207]: done

File: ydb/core/tx/schemeshard/schemeshard__login.cpp (hunk 6 above)
Comment (molotkov-and) [reply ydb/core/tx/schemeshard/schemeshard__login.cpp:111]: Cannot test code returns this error. I think this piece of code is unreachable

File: ydb/core/tx/schemeshard/schemeshard_impl.cpp (hunk 7 above)
Comment (molotkov-and) [reply ydb/core/tx/schemeshard/schemeshard_impl.cpp:4425]: I think we cannot test this case. It is unclear how many login attempts need to be done

Comment (azevaykin) [reply ydb/core/tx/schemeshard/schemeshard_impl.cpp:4425]: May be something like `runtime->AdvanceCurrentTime(TDuration::Years(1));` ?

File: ydb/core/tx/schemeshard/schemeshard__login.cpp [hunk 8]
```diff
@@ -67,37 +69,133 @@ struct TSchemeShard::TTxLogin : TSchemeShard::TRwTxBase {
...
+                Result->Record.SetError(TStringBuilder() << "User " << loginRequest.User << " is locked out");
+                return true;
             }
+        } else if (ShouldResetFailedAttemptCount(lastFailedAttempt)) {
+            ResetFailedAttemptCount(loginRequest, db);
+        }
+        const NLogin::TLoginProvider::TLoginUserResponse loginResponse = Self->LoginProvider.LoginUser(loginRequest);
+        switch (loginResponse.Status) {
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::SUCCESS: {
+            HandleLoginAuthSuccess(loginRequest, loginResponse, db);
```
Comment (ijon) [inline ydb/core/tx/schemeshard/schemeshard__login.cpp:149]: `HandleLoginAuthSuccess()` and `HandleLoginAuthInvalidPassword()` both hide how the final `Result` is set. While the whole point of `HandleLoginAuth()` is to construct that `Result`. So hiding it does not help the reader to understand how `Result` is constructed.

I would suggest moving `Result` setting out of those functions and back to `HandleLoginAuth()`.

**Review by ijon — DISMISSED** [review]

File: ydb/tests/library/common/types.py [hunk 9]
```diff
@@ -248,6 +248,8 @@ class PType(AbstractTypeEnum):
     Double = _ptype_from(32, float_in(-100, 100), float, proto_field='Double')
     Float = _ptype_from(33, float_in(-100, 100), float, proto_field='Float')
 
+    Timestamp = _ptype_from(50, int_between(0, 2 ** 64 - 1), int, proto_field='Timestamp', min_value=0, max_value=2 ** 64 - 1)
```
[the remark this answers is not in the document]
Comment (molotkov-and) [reply ydb/tests/library/common/types.py:251]: Yes, it does. I added to schemeshard localDB new columns with type `Timestamp` and errors appeared in functional tests. `Do not known type with id 50` That is just `Timestamp` which i added

File: ydb/core/tx/schemeshard/schemeshard__login.cpp (hunk 8 above)
Comment (molotkov-and) [reply ydb/core/tx/schemeshard/schemeshard__login.cpp:145]: done

**Review by azevaykin — APPROVED** [review]

**Pull request merged**